slateratom/lib/avgpot.f90

```diff
@@ -0,0 +1,450 @@
+!> Module to calculate the approximate average potential and effective orbital energies proposed in:
+!! Baer R, Livshits E, Salzner U. Tuned range-separated hybrids in density functional theory.
+!! Annual review of physical chemistry. 2010 May 5;61:85-109.
+module average_potential
+
+  use common_accuracy, only : dp, mc
+  use common_constants, only : pi
+  use common_message, only : error
+
+  use utilities, only : index_heap_sort
+  use density, only : wavefunction, wavefunction_1st, wavefunction_2nd
+
+  implicit none
+  private
+
+  public :: getAveragePotential
+
+
+contains
+
+
+  !> Tries to infer the energy of the highest occupied atomic orbital (HOAO) and the associated
+  !! principal and angular quantum number from the eigenvalues and occupations handed over.
+  !!
+  !! The effort made here would be an argument for moving this functionality to the sktools
+  !! sometime in the future.
+  subroutine getHoaoOrLowestNl(eigval, occ, max_l, num_alpha, poly_order, hoaoN, hoaoL, eHoao)
+
+    !> Eigenvalues of selected spin-channel
+    real(dp), intent(in) :: eigval(0:, :)
+
+    !> Occupation numbers of selected spin-channel
+    real(dp), intent(in) :: occ(0:, :)
+
+    !> Maximum angular momentum
+    integer, intent(in) :: max_l
+
+    !> Number of exponents in each shell
+    integer, intent(in) :: num_alpha(0:)
+
+    !> Highest polynomial order + l in each shell
+    integer, intent(in) :: poly_order(0:)
+
+    !> Principal and angular quantum number of HOAO eigenvalue
+    integer, intent(out) :: hoaoN, hoaoL
+
+    !> Energy of highest occupied atomic orbital
+    real(dp), intent(out) :: eHoao
+
+    !! Flattened, one-dimensional versions of occupations and eigenvalues
+    real(dp), allocatable :: occFlat(:), eigvalFlat(:)
+
+    !! Index array that sorts the flattened eigenvalues
+    integer, allocatable :: eigvalFlatSortInds(:)
+
+    !! True, if all occupations of current spin channel are below threshold
+    logical :: isUnoccupied
+
+    !! Auxiliary variables
+    integer :: ll, nn, iDummy1, iDummy2, iHoao, ii, ind
+
+    eigvalFlat = reshape(eigval, [size(eigval)])
+    occFlat = reshape(occ, [size(occ)])
+
+    allocate(eigvalFlatSortInds(size(eigvalFlat)))
+    call index_heap_sort(eigvalFlatSortInds, eigvalFlat)
```

Comment:
I just switched over to the brute-force search suggested by you, which also makes adding sorting routines to the utilities unnecessary. As a side-effect the degeneracy bug pointed out by you should be resolved, since the energy criterion to search for the corresponding (nn,ll) tuple is supplemented by checking the occupation simultaneously, in contrast to the previous version.

Reply:
@bhourahine I will mark it as resolved for now, but please have a look at it again in the next iteration.